pycqed/measurement/hdf5_data.py

```diff
@@ -124,9 +125,150 @@ def encode_to_utf8(s):
+            write_dict_to_hdf5(data_dict=item,
+                               entry_point=entry_point[key])
+        elif isinstance(item, (list, tuple)):
+            if len(item) > 0:
+                elt_type = type(item[0])
+                # Lists of a single type, are stored as an hdf5 dset
+                if (all(isinstance(x, elt_type) for x in item) and
+                        not isinstance(item[0], dict) and
+                        not isinstance(item, tuple)):
+                    if isinstance(item[0], (int, float,
+                                            np.int32, np.int64)):
+                        entry_point.create_dataset(key,
+                                                   data=np.array(item))
+                        entry_point[key].attrs['list_type'] = 'array'
+
+                    # strings are saved as a special dtype hdf5 dataset
+                    elif isinstance(item[0], str):
+                        dt = h5py.special_dtype(vlen=str)
+                        data = np.array(item)
+                        data = data.reshape((-1, 1))
+                        ds = entry_point.create_dataset(
+                            key, (len(data), 1), dtype=dt)
+                        ds.attrs['list_type'] = 'str'
+                        ds[:] = data
+                    else:
+                        logging.warning(
+                            'List of type "{}" for "{}":"{}" not '
+                            'supported, storing as string'.format(
+                                elt_type, key, item))
+                        entry_point.attrs[key] = str(item)
+                # Storing of generic lists/tuples
+                else:
+                    entry_point.create_group(key)
+                    # N.B. item is of type list
+                    list_dct = {'list_idx_{}'.format(idx): entry for
+                                idx, entry in enumerate(item)}
+                    print(list_dct)
+                    group_attrs = entry_point[key].attrs
+                    if isinstance(item, tuple):
+                        group_attrs['list_type'] = 'generic_tuple'
+                    else:
+                        group_attrs['list_type'] = 'generic_list'
+                    group_attrs['list_length'] = len(item)
+                    write_dict_to_hdf5(
+                        data_dict=list_dct,
+                        entry_point=entry_point[key])
+            else:
+                # as h5py does not support saving None as attribute
+                entry_point.attrs[key] = 'NoneType:__emptylist__'
+
+        else:
+            logging.warning(
+                'Type "{}" for "{}" (key): "{}" (item) at location {} '
+                'not supported, '
+                'storing as string'.format(type(item), key, item,
+                                           entry_point))
+            entry_point.attrs[key] = str(item)
+
+
+def read_dict_from_hdf5(data_dict: dict, h5_group):
```

Comment:
Why is the first argument needed? Wouldn't it be better to create a new dictionary in this function and then just return it. Then the user calls

`loaded_dict = read_dict_from_hdf5(my_h5_group)`

instead of

`loaded_dict = {}
read_dict_from_hdf5(loaded_dict, my_h5_group)`